protocol/l2-forked-tests.md

```diff
@@ -0,0 +1,209 @@
+
+This framework should be able to take a network's initial state and apply the required Network Upgrade Transactions (NUTs) to reach the latest version of the contracts for running the test suite. This approach would help increase confidence that the upgrade process doesn't introduce unexpected bugs, allowing contributors to catch errors earlier in the release process. While this implementation won't perfectly mirror the environment after NUTs are applied to the real network, it represents a significant improvement to our testing capabilities.
+
+The solution must provide a deterministic state to the test suite, enabling the execution of targeted test cases against this final state. This state results from applying a set of NUTs to the initial forked state. Therefore, we need to implement a structured way to describe these NUTs and a method to apply them.
+
+It's also important that this solution remains loosely coupled with the test setup itself, so that running tests against different chains starting with different initial states requires minimal changes to the Setup test file.
+
+## Proposed Solution
+
+Currently, L1 forked state is abstracted by ForkLive, which handles reading the L1 contract's code and upgrading when needed. Ideally, we'd like a similar approach for L2 fork tests, where a single contract is responsible for setting up L2 Predeploys in the final state needed for our test suite. We propose a two-part approach: one part abstracts the NUTs that occur in each upgrade, while the other part executes them sequentially. For L2, we have the advantage of knowing the addresses in advance, so unlike ForkLive, storing addresses isn't necessary.
+
+### Implementation Components
+
+1. **NUT Definitions**: Specifications of the Network Upgrade Tests to be executed prior to test suite initialization
+
+```solidity
+interface NUTExecutor {
+    function execute(bytes calldata _calldata) external;
+}
+
+contract XForkExecutor is NUTExecutor {
+    function execute(bytes calldata _calldata) external {
+        ///         1. Deploy a new `L1BlockImpl` contract.
+        ///         2. Upgrade only the `L1Block` contract to the new implementation by
+        ///            calling `L2ProxyAdmin.upgrade(address(L1BlockProxy), address(L1BlockImpl))`.
+        ///         3. Call `L1Block.setXFork()` to pull the values from L2 contracts.
+        ///         4. Upgrades the remainder of the L2 contracts via `L2ProxyAdmin.upgrade()`.
+    }
+}
+
+```
+
+1. **Execution Script**: A specialized module that manages the sequential execution of defined NUTs
+
+```solidity
+contract Upgrader {
+    function upgrade(NUTExecutor[] memory _executors) external {
```

Comment:
The idea here is that you can apply multiple upgrades, similar to how Forklive calls `opcm.upgrade()` on multiple OPCM versions?